vowpalwabbit/object_pool.h

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include <set>
+#include <queue>
+
+namespace VW
+{
+template <typename T>
+struct object_pool
+{
+  virtual void return_object(T*) = 0;
+  virtual T* get_object() = 0;
+  virtual bool empty() const = 0;
+  virtual size_t available() const = 0;
+  virtual size_t size() const = 0;
+  virtual bool is_from_pool(T*) const = 0;
+
+  virtual ~object_pool() {}
+};
+
+// Not thread safe
+template <typename T, typename TFactory>
+struct unbounded_object_pool : object_pool<T>
+{
+  unbounded_object_pool() = default;
+  unbounded_object_pool(size_t initial_size)
+  {
+    for (size_t i = 0; i < initial_size; i++)
+    {
+      auto obj = m_factory();
```

Comment:
We should allocate this initial set of objects through `new T[]` to make sure they are contiguous and packed.

Reply:
Is it possible to mix the initial block initialization with the normal factory usage in a generic way that encompasses both? I agree that contiguous and packed would be optimal but not 100% sure how to get it to work

Reply:
If we allocate in chunks we can test membership by bounds checking against them.
We can have a fast path with data inline in the object pool itself with overflow using a std::vector of tuples.

I'll leave as an exercise to check when the constant factor cut in the linear cost due to chunking matches the cost of the logarithmic lookup of std::set.

Reply:
Can you explain more what you mean by a fast path with data inline using tuples?

We can get away from log lookup in the set by using an unordered set, and since we are storing pointers in it the hash is trivial